### PR TITLE
hub: Added multitopic to UMB

### PR DESCRIPTION
**Pull request by jperezdealgaba**
Resolves: https://issues.redhat.com/browse/OSH-215

Added the logic to send UMB messages to multiple topics at the same time.

### CODE REVIEW

File: osh/hub/scan/messaging.py
Comment (kdudka) [inline osh/hub/scan/messaging.py:29]: I think this should be named `topic_prefix` for consistency (and to make it obvious that it will be extended).

File: osh/hub/scan/messaging.py
Comment (kdudka) [inline osh/hub/scan/messaging.py:75]: This needs to be implemented in a backward-compatible way so that updated code works with existing configuration.  I would keep the name `UMB_TOPIC_PREFIX` and just extend its semantics such that it can be either a string or a list of strings.  Apart from backward-compatibility, an advantage is that you do not need to document what will happen when both `UMB_TOPIC_PREFIX` and `UMB_TOPIC_PREFIX_LIST` are set.

**Review by kdudka — CHANGES_REQUESTED** [review]

File: osh/hub/scan/messaging.py
Comment (jperezdealgaba) [reply osh/hub/scan/messaging.py]: Just changed it!

File: osh/hub/scan/messaging.py
Comment (jperezdealgaba) [reply osh/hub/scan/messaging.py]: I modified the semantics so it is backwad-compatible with existing configuration.

Let me know please if that is what you meant. 

Thank you!

Comment (lzaoral) [reply osh/hub/scan/messaging.py]: @kdudka I know that you want your upgrades as seamless as possible but I'd like to see this hack removed after some time in the future.  The code of OSH is already excessively convoluted as it is.

**Review by lzaoral — CHANGES_REQUESTED** [review]
See comments.  Also, the PR and commit messages do not adhere to the contribution guidelines.

File: osh/hub/settings_local.py
Comment (jperezdealgaba) [reply osh/hub/settings_local.py]: That was introduced by the linter. Fixed

File: osh/hub/settings_local.py
Comment (jperezdealgaba) [reply osh/hub/settings_local.py]: To avoid this confusion, I rolled back this setting to the single string. The internal configuration will make use of the list

File: osh/hub/settings_local.py
Comment (lzaoral) [reply osh/hub/settings_local.py]: Hmm, that it weird.  `pre-commit` should not reformat any lines that were not edited.

Comment (jperezdealgaba) [reply osh/hub/settings_local.py]: On PyCharm, if you click on "Reformat the file" to fix some suggestion, it will automatically format the whole document, also even for lines that were not edited.
Screenshot 2024-01-24 at 13 56 10

File: osh/hub/settings_local.py
Comment (lzaoral) [reply osh/hub/settings_local.py]: Using a list is completely fine and reasonable.  My only concern was the unsolicited modification of the actual topic prefixes.

File: osh/hub/scan/messaging.py
Comment (lzaoral) [reply osh/hub/scan/messaging.py:75]: @kdudka Yes, that's better and it also allows other collections like `dict` or `tuple` to be used to specify topic prefixes.

File: osh/hub/scan/messaging.py
Comment (jperezdealgaba) [reply osh/hub/scan/messaging.py:58]: @kdudka Shouldn't it here be `topic_prefix` as it is not yet the whole topic and at the `UMBSender` be `topic` when it is concatenated with the `key`?

Comment (kdudka) [reply osh/hub/scan/messaging.py:58]: Indeed.  It is the other way around: `self.topic` in `SenderThread` needs to be renamed to `self.topic_prefix`.  I confused it with `UMBSender`, which had already been using `self.topic` but correctly (for the whole topic).

File: osh/hub/scan/messaging.py
Comment (jperezdealgaba) [reply osh/hub/scan/messaging.py]: I have used your code and also have documented it in the settings so it is clear what can be used. Once this approved, I will update the internal config file.

**Comment by hanchuntao** [comment]
Hello @jperezdealgaba . I don't know if we need to update `openscanhub/osh/hub/scripts/umb-emit.py`

**Comment by jperezdealgaba** [comment]
@hanchuntao GitHub doesn't let me answer your comment directly. I don't see the need as `openscanhub/osh/hub/scripts/umb-emit.py` it is a scrip to test the UMB connection and the only thing changing here is the topic (which was already changed to `openscanhub` before I started this) and the certificate name, but we may use another one.

Let me know if you think I should do it.

**Comment by kdudka** [comment]
Yes, the `umb-emit.py` script is mainly for debugging and it does not need to send messages to multiple topics at a time.

File: osh/hub/scan/messaging.py
Comment (kdudka) [reply osh/hub/scan/messaging.py]: @jperezdealgaba Why did you discard the empty line and comments from my proposal?  They were there to improve code readability...

Comment (jperezdealgaba) [reply osh/hub/scan/messaging.py]: I thought it was for explanation. Let me add them again

Comment (kdudka) [reply osh/hub/scan/messaging.py]: Yes, please.  The explanation might be pretty useful for anyone who will be reading the code later on.

Comment (jperezdealgaba) [reply osh/hub/scan/messaging.py]: Added

**Review by kdudka — APPROVED** [review]
Looks good.  Thanks for the update!

File: osh/hub/settings_local.ci.py
Comment (lzaoral) [inline osh/hub/settings_local.ci.py:115]:
```suggestion
# UMB_TOPIC_PREFIX can be a single string or a collection of strings.
# Every UMB message will be sent to all defined UMB topic prefixes.
```
* The code you committed will also work for other collections like tuples or sets.  Besides, the second sentence was incomplete.
* I also suggest to enable a spell checker in your editor for future contributions.

File: osh/hub/settings_local.py
Comment (lzaoral) [inline osh/hub/settings_local.py:125]: See the first comment.

**Review by lzaoral — CHANGES_REQUESTED** [review]
Final remarks.  I'll approve the PR to be merged when they are successfully incorporated.

File: osh/hub/settings_local.py
Comment (jperezdealgaba) [reply osh/hub/settings_local.py]: Fixed

File: osh/hub/settings_local.ci.py
Comment (jperezdealgaba) [reply osh/hub/settings_local.ci.py]: Fixed and agree.

**Review by lzaoral — APPROVED** [review]

**Comment by lzaoral** [comment]
@jperezdealgaba I've just noticed that the issue linked in the `Resolves:` clause in the commit message is unrelated and different than the one in the PR description.  Could you please fix that as well?  Thank you!

**Comment by kdudka** [comment]
Thank you both!  Merging...

**Pull request merged**
